### PR TITLE
Allow for hdfs and gcs URI's to be passed to GenomicsDB

### PR DESCRIPTION
**Pull request by droazen**
This PR is a finalized version of https://github.com/broadinstitute/gatk/pull/5017. I've copied the branch into our repo so that travis will run cloud tests on it.

Currently, only Posix filesystem paths can be passed as workspaces and arrays to GenomicsDB via GenomicsDBImport and SelectVariants. This PR will allow for hdfs and gcs (and emrfs/s3) URIs to be supported as well. 
Examples
   ```
./gatk GenomicsDBImport -V /vcfs/sample.vcf.gz --genomicsdb-workspace-path hdfs://master:9000/gdb_ws  -L 1:500-10000
   ./gatk GenomicsDBImport -V /vcfs/sample.vcf.gz --genomicsdb-workspace-path gs://my_bucket/gdb_ws  -L 1:500-10000
```
```
  ./gatk SelectVariants -V gendb.hdfs://master:9000/gdb_ws -R hs37d5.fa -O out.vcf
 ./gatk SelectVariants -V gendb.gs://my_bucket/gdb_ws -R hs37d5.fa -O out.vcf
```
GenomicsDB supports GCS via the [Cloud Storage Connector](https://cloud.google.com/dataproc/docs/concepts/connectors/cloud-storage). Set environment variable GOOGLE_APPLICATION_CREDENTIALS to point to the GCS Service Account json file.

### CODE REVIEW

**Comment by droazen** [comment]
@nalinigans There are some test failures in the GenomicsDB cloud tests. See: https://storage.googleapis.com/hellbender-test-logs/build_reports/master_22153.1/tests/test/index.html

We'll have to fix these before we can merge.

**Comment by nalinigans** [comment]
@droazen, will put some debug print statements in the two tests that are failing while authenticating with GCS and issue another pull request to _nalinigans_genomicsdb_uri_support_ branch. Hope that is OK. Thanks.

**Comment by nalinigans** [comment]
@droazen, we have a free GCS account, so it is possible that Hadoop requires extra configuration. Can anyone help here? The is the code we have for connecting to GCS via Hadoop.

```
hdfsFS gcs_connect(struct hdfsBuilder *builder, const std::string& working_dir) {
  char *gcs_creds = getenv("GOOGLE_APPLICATION_CREDENTIALS");
  if (gcs_creds) {
    value = parse_json(gcs_creds, "project_id"); // free value after hdfsBuilderConnect as it is shallow copied.
    if (value) {
      hdfsBuilderConfSetStr(builder, "google.cloud.auth.service.account.enable", "true");
      hdfsBuilderConfSetStr(builder, "google.cloud.auth.service.account.json.keyfile", gcs_creds);
      hdfsBuilderConfSetStr(builder, "fs.gs.project.id", value);
    }
  }

  if (working_dir.empty()) {
    hdfsBuilderConfSetStr(builder, "fs.gs.working.dir", "/");
  } else {
    hdfsBuilderConfSetStr(builder, "fs.gs.working.dir", working_dir.c_str());
  }

  // Default buffer sizes are huge in the GCS connector. GenomicsDB reads/writes in smaller chunks,
  // so the buffer size can be made a little smaller.
  hdfsBuilderConfSetStr(builder, "fs.gs.io.buffersize.write", "262144");

  hdfsFS hdfs_handle = hdfsBuilderConnect(builder);
  free(value);
  return hdfs_handle;
}
```

This is the error from Travis logs-
```
Running Test: Test method testWriteToAndQueryFromGCS(org.broadinstitute.hellbender.tools.genomicsdb.GenomicsDBImportIntegrationTest)
hdfsBuilderConnect(forceNewInstance=1, nn=gs://hellbender-test-logs, port=0, kerbTicketCachePath=(NULL), userName=(NULL)) error:
java.io.IOException: Error getting access token from metadata server at: http://metadata/computeMetadata/v1/instance/service-accounts/default/token
        at com.google.cloud.hadoop.util.CredentialFactory.getCredentialFromMetadataServiceAccount(CredentialFactory.java:210)
        at com.google.cloud.hadoop.util.CredentialConfiguration.getCredential(CredentialConfiguration.java:75)
        at com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystemBase.configure(GoogleHadoopFileSystemBase.java:1826)
  Caused by: com.google.api.client.http.HttpResponseException: 404 Not Found
{"error":"invalid_request","error_description":"Service account not enabled on this instance"}
        at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1072)
        at com.google.cloud.hadoop.util.CredentialFactory$ComputeCredentialWithRetry.executeRefreshToken(CredentialFactory.java:159)
        at com.google.api.client.auth.oauth2.Credential.refreshToken(Credential.java:493)
        at com.google.cloud.hadoop.util.CredentialFactory.getCredentialFromMetadataServiceAccount(CredentialFactory.java:208)
        ... 77 more
[GenomicsDB::StorageManagerConfig] Error: Error getting hdfs connection: Connection refused.
```

**Comment by nalinigans** [comment]
@droazen , I got a genomicsdb.jar from @kgururaj  and just tried out the GenomicsDB cloud tests. The call stack that I got from my test run in nalini_new_genomicsdb_jar mentions that we do need the **fs.gs.project.id** hadoop configuration set. The service json I use for our internal testing has this key, but the Hellbender service json does not. Any ideas on how to get this key for the tests? How is it being made available to the spark cloud tests for example? I do see it being configured in src/main/java/org/broadinstitute/hellbender/engine/spark/SparkContextFactory.java.

```
hdfsBuilderConnect(forceNewInstance=1, nn=gs://hellbender-test-logs, port=0, kerbTicketCachePath=(NULL), userName=(NULL)) error:
java.io.IOException: Must supply a value for configuration setting: **fs.gs.project.id**
	at com.google.cloud.hadoop.util.ConfigurationUtil.getMandatoryConfig(ConfigurationUtil.java:39)
	at com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystemBase.createOptionsBuilderFromConfig(GoogleHadoopFileSystemBase.java:2185)
	at com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystemBase.configure(GoogleHadoopFileSystemBase.java:1832)
	at com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystemBase.initialize(GoogleHadoopFileSystemBase.java:1013)
	at com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystemBase.initialize(GoogleHadoopFileSystemBase.java:976)
```

**Comment by nalinigans** [comment]
PR is ready to be merged. The 2 GCS tests in GenomicsDBImportIntegrationTest.java are commented out, but they have been tested with the HELLBENDER test project and GOOGLE_APPLICATION_CREDENTIALS in the nalini_new_genomicsdb_jar branch and will be uncommented as soon as a new GenomicsDB jar is released.

**Comment by lbergelson** [comment]
#5305

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 1]
```diff
@@ -712,26 +711,20 @@ public VariantContext next() {
...
+    private String overwriteOrCreateWorkspace() {
+        String workspaceDir = BucketUtils.makeFilePathAbsolute(workspace);
+        // From JavaDoc for GenomicsDBUtils.createTileDBWorkspace
+        //   returnCode = 0 : OK. If overwriteExistingWorkspace is true and the workspace exists, it is deleted first.
+        //   returnCode = -1 : path was not a directory
+        //   returnCode = -2 : failed to create workspace
+        //   returnCode = 1 : if overwriteExistingWorkspace is false, return 1 if directory already exists
+        int returnCode = GenomicsDBUtils.createTileDBWorkspace(workspaceDir, overwriteExistingWorkspace);
+        if (returnCode < 0) {
+            throw new UnableToCreateGenomicsDBWorkspace("Error creating GenomicsDB workspace: " + workspace);
```
Comment (ldgauthier) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: Can we have a different exception for returnCode -1 noting that the path was not a directory?  That seems like an easier error to address than whatever caused -2.

File: src/main/java/org/broadinstitute/hellbender/utils/io/IOUtils.java [hunk 2]
```diff
@@ -864,4 +906,85 @@ public static final String urlDecode(final String string) {
...
+        return getGenomicsDBPath(path) != null;
+    }
+
+    /**
+     * Get the GenomicsDB equivalent absolute URL for a given path
+     *
+     * @param gendbPath String representing legal gendb URI
+     * @return absolute gendb URI to the path
+     */
+    public static String getAbsolutePathWithGenDBScheme(final String gendbPath) {
```
Comment (ldgauthier) [inline src/main/java/org/broadinstitute/hellbender/utils/io/IOUtils.java]: I'd rather not abbreviate GenomicsDB in this method name. Maybe "getAbsolutePathWithGenomicsDBURIScheme"?

Comment (nalinigans) [reply src/main/java/org/broadinstitute/hellbender/utils/io/IOUtils.java]: Done. See [commit](url
https://github.com/broadinstitute/gatk/commit/6c860bd4559e3b2f3af612fe95d5c0efa4b6a391).

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 1 above)
Comment (nalinigans) [reply src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: Done. See [commit](https://github.com/broadinstitute/gatk/commit/6c860bd4559e3b2f3af612fe95d5c0efa4b6a391).

**Comment by droazen** [comment]
@lbergelson is going to try regenerating our service account key and see if that allows us to uncomment the GCS tests in this branch.

**Comment by nalinigans** [comment]
@lbergelson, thanks for regenerating the service account key. All tests pass now. 
@droazen, feel free to merge with master.

File: src/main/java/org/broadinstitute/hellbender/engine/FeatureDataSource.java [hunk 3]
```diff
@@ -350,36 +350,22 @@ public FeatureDataSource(final FeatureInput<T> featureInput, final int queryLook
...
-        final File workspace = new File(noheader);
-        final File callsetJson = new File(noheader, GenomicsDBConstants.DEFAULT_CALLSETMAP_FILE_NAME);
-        final File vidmapJson = new File(noheader, GenomicsDBConstants.DEFAULT_VIDMAP_FILE_NAME);
-        final File vcfHeader = new File(noheader, GenomicsDBConstants.DEFAULT_VCFHEADER_FILE_NAME);
-
-        if (!workspace.exists() || !workspace.canRead() || !workspace.isDirectory()) {
-            throw new UserException("GenomicsDB workspace " + workspace.getAbsolutePath() + " does not exist, " +
-                    " is not readable, or is not a directory");
+        final String workspace = IOUtils.getGenomicsDBAbsolutePath(path) ;
+        if (workspace == null) {    throw new IllegalArgumentException("Trying to create a GenomicsDBReader from  non-GenomicsDB inputpath " + path);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/FeatureDataSource.java]:
```suggestion
        if (workspace == null) {   
            throw new IllegalArgumentException("Trying to create a GenomicsDBReader from  non-GenomicsDB inputpath " + path);
```

File: src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java [hunk 4]
```diff
@@ -1,5 +1,8 @@
 package org.broadinstitute.hellbender.tools.genomicsdb;
 
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java]:
```suggestion
```

File: src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java [hunk 5]
```diff
@@ -1,5 +1,8 @@
 package org.broadinstitute.hellbender.tools.genomicsdb;
 
+import com.fasterxml.jackson.core.type.TypeReference;
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java]:
```suggestion
```

File: src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java [hunk 6]
```diff
@@ -802,26 +801,26 @@ public void testPreserveContigOrderingInHeader() throws IOException {
             final String workspace, final String reference,
             final boolean produceGTField,
             final boolean sitesOnlyQuery) throws IOException {
-        GenomicsDBExportConfiguration.ExportConfiguration.Builder exportConfigurationBuilder = GenomicsDBExportConfiguration.ExportConfiguration.newBuilder()
+       String workspaceAbsPath = BucketUtils.makeFilePathAbsolute(workspace);GenomicsDBExportConfiguration.ExportConfiguration .Builder exportConfigurationBuilder = GenomicsDBExportConfiguration.ExportConfiguration.newBuilder()
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java]:
```suggestion
       String workspaceAbsPath = BucketUtils.makeFilePathAbsolute(workspace);
       GenomicsDBExportConfiguration.ExportConfiguration.Builder exportConfigurationBuilder = GenomicsDBExportConfiguration.ExportConfiguration.newBuilder()
```

**Comment by droazen** [comment]
@nalinigans Tests passed, but it's worth pointing out that the new `testWriteToAndQueryFromGCS()` test took 9 minutes to complete, which seems very slow. Possible performance issue?

**Pull request merged**